gobblin-runtime/src/main/java/gobblin/runtime/JobContext.java

```diff
@@ -267,6 +272,13 @@ void storeJobExecutionInfo() {
     }
   }
 
+  @Subscribe
+  public void handleNewOutputTaskStateEvent(NewOutputTaskStateEvent newOutputTaskStateEvent) {
+    this.jobState.addTaskState(newOutputTaskStateEvent.getTaskState());
+    // Update the job execution history store upon every task completion
+    storeJobExecutionInfo();
```

Comment:
The overhead is pretty much negligible compared with the cost of collecting the `TaskState`. But yes we can definitely change this to carry a list of `TaskState`s.